pymc3/ode/ode.py

```diff
@@ -95,89 +102,104 @@ def _make_sens_ic(self):
+        y = sol[1:, :self.n_states]
 
         # The sensitivities, reshaped to be a sequence of matrices
-        sens = sol[1:, self.n_states :].reshape(len(self.times), self._n, self._m)
+        sens = sol[1:, self.n_states:].reshape(self.n_times, self.n_states, self.n_p)
 
         return y, sens
 
-    def _cached_simulate(self, parameters):
-        if np.array_equal(np.array(parameters), self._cached_parameters):
-
-            return self._cached_y, self._cached_sens
-
-        return self._simulate(np.array(parameters))
-
-    def _state(self, parameters):
-        y, sens = self._cached_simulate(np.array(parameters))
-        self._cached_y, self._cached_sens, self._cached_parameters = y, sens, parameters
-        return y.ravel()
-
-    def _numpy_vsp(self, parameters, g):
-        _, sens = self._cached_simulate(np.array(parameters))
-
-        # Each element of sens is an nxm sensitivity matrix
-        # There is one sensitivity matrix per time step, making sens a (len(times), n_states, len(parameter))
-        # dimensional array.  Reshaping the sens array in this way is like stacking each of the elements of sens on top
-        # of one another.
-        numpy_sens = sens.reshape((self.n_states * len(self.times), len(parameters)))
-        # The dot product here is equivalent to np.einsum('ijk,jk', sens, g)
-        # if sens was not reshaped and if g had the same shape as yobs
-        return numpy_sens.T.dot(g)
-
-    def make_node(self, odeparams, y0):
-        if len(odeparams) != self.n_odeparams:
-            raise ValueError(
-                "odeparams has too many or too few parameters.  Expected {a} parameter(s) but got {b}".format(
-                    a=self.n_odeparams, b=len(odeparams)
-                )
-            )
-        if len(y0) != self.n_states:
-            raise ValueError(
-                "y0 has too many or too few parameters.  Expected {a} parameter(s) but got {b}".format(
-                    a=self.n_states, b=len(y0)
-                )
+    def make_node(self, y0, theta):
+        inputs = (y0, theta)
+        _log.debug('make_node for inputs {}'.format(hash(inputs)))
+        states = self._otypes[0]()
+        sens = self._otypes[1]()
+
+        # store symbolic output in dictionary such that it can be accessed in the grad method
+        self._output_sensitivities[hash(inputs)] = sens
+        return theano.Apply(self, inputs, (states, sens))
+
+    def __call__(self, y0, theta, return_sens=False, **kwargs):
+        # convert inputs to tensors (and check their types)
+        y0 = tt.cast(tt.unbroadcast(tt.as_tensor_variable(y0), 0), theano.config.floatX)
+        theta = tt.cast(tt.unbroadcast(tt.as_tensor_variable(theta), 0), theano.config.floatX)
+        inputs = [y0, theta]
+        for i, (input, itype) in enumerate(zip(inputs, self._itypes)):
+            assert input.type == itype, 'Input {} of type {} does not have the expected type of {}'.format(i, input.type, itype)

```

Comment:
this is not so important, but inline `assert`s for exceptions are antipatterns: you can disable them via a flag when you run python (i think `-o`). These could be swapped out for a `ValueError`, `RuntimeError`, or (probably best) a custom error.

For now, maybe at least a `TODO` that marks this as useful work, since I have never met anyone who uses that flag.

Reply:
Excellent point. I've switched to `ValueError` and a new custom `ShapeError`.
This way, the coverage also picks up that they are currently not tested.